sherpa/astro/data.py

```diff
@@ -1523,23 +1523,44 @@ def _get_rate(self):
     def _set_rate(self, val):
         self._rate = bool_cast(val)
         for id in self.background_ids:
-            # TODO: shouldn't this store bool_cast(val) instead?
             self.get_background(id).rate = val
 
     rate = property(_get_rate, _set_rate,
-                    doc='Quantity of y-axis: counts or counts/sec')
+                    doc="""Is the Y axis displayed as a rate when plotting data?
+
+When True the y axis is normalised by the exposure time to display
+a rate.""")
 
     def _get_plot_fac(self):
         return self._plot_fac
 
     def _set_plot_fac(self, val):
-        self._plot_fac = int(val)
+        # I'd prefer to check whether val is an integer, but there may
+        # be users who have set the value to 2.0 and it doesn't seem
+        # worth breaking that code. We do however want to error out if
+        # val=0.5 (say), hence this slightly-contrived check.
+        #
+        try:
+            ival = int(val)
+            okay = val == ival
+        except (ValueError, TypeError):
+            # For when int() can't convert val, which can raise
+            # different errors.
+            okay = False
+
+        if not okay:
+            raise DataErr("bad", "plot_fac setting", val)
```

Comment:
I think I still see a problem here, but I need to play around with the old version more to see what the differences are. Unfortunately, I might not get to that today.

Reply:
There's no hurry.

Reply:
If we didn't want to try and be backwards compatible - i.e. we don't need to support "set_analysis(... , factor=2.0)" - then it would be easier. I note in matplotlib 5.3.0 that they have a bunch of changes like this (I think), but I'm not sure our user community has the same setup as they do.

I'm not particularly happy with the code, but it works (as far as I can tell).

Reply:
What exactly is this supposed to accept?
- Do inputs have to be integers, i.e. is there a good physical reason to allow `factor=2`, but not `factor=3/2`? That asks for what scalings make sense. Surely 0 and 1 do make sense, but why not 0.5? I can imagine someone scaling with sqrt(E) or sqrt(lambda), in which case such a plot would make sense or not?
- What inputs do we want to accept? Int, float, also string? I can only trigger the except clause with exotic cases like strings, complex number etc. If the goal of the except clause is to give a better error message for strings, etc., I would suggest:
```
except (ValueError, TypeError):
    raise ArgumentTypeErr("badarg", val, "int or float")
```

Code that I used to try to understand what happens
```
def _set_plot_fac(val):
    '''removed "self" argument and turned output and error into print statements
    so I can call it a couple of times in a row'''
    try:
        ival = int(val)
        okay = val == ival
    except (ValueError, TypeError):
        # For when int() can't convert val, which can raise
        # different errors.
        print("except clause triggered")
        okay = False

    if not okay:
        print(f"{val=} not okay")
        return
            
    print(ival)


0
1
2
val=2.5 not okay
0
1
2
val=2.5 not okay
val='0' not okay
val='1' not okay
val='2' not okay
except clause triggered
val='2.5' not okay
except clause triggered
val='0.' not okay
except clause triggered
val='1.' not okay
except clause triggered
val='2.' not okay
except clause triggered
val='2.5' not okay
```

Reply:
It accepts an integer. There is no support for non-integer values. adding support for factor=0.5 would be a much-larger PR. This is only about improving the existing support.

When setting plot_fac - either directly as an attribute or via the factor argument of set_analysis - the original code would take the value you send it (`val`) and then call int on it (`int(val)`) - so that if you sent it 0.5 it would silelntly get converted to 0. All I'm really trying to do is ensure we test these assumptions.

So your output above looks like I'd expect it to.

The one question is that you used to be able to set plot_fac to "1" or "2" and this is no-longer supported, but I am pretty sure we never documented support for this. There are a number of places in our code base hwere we can accept strings rather than numbers which comes from "pre history" (e.g. when supporting S-Lang as well as Python and a spill over from the old command language for Sherpa).